bluemira/equilibria/coils/_grouping.py

```diff
@@ -1015,3 +1157,287 @@ def from_group_vecs(
+        ]
+        return [*flatten_iterable(optimisable_coil_names)]
+
+    @property
+    def all_current_optimisable_coils(self) -> list[Coil]:
+        """
+        Get the names of all coils that can be current optimised.
+        """
+        return [self[cn] for cn in self.current_optimisable_coil_names]
+
+    def get_current_optimisable_coils(
+        self, coil_names: list[str] | None = None
+    ) -> list[Coil]:
+        """
+        Get the coils that can be current optimised.
+        """
+        if coil_names is None:
+            return self.all_current_optimisable_coils
+
+        opt_coils_map = {c.name: c for c in self.all_current_optimisable_coils}
+        rtn = []
+        for cn in coil_names:
+            c = opt_coils_map.get(cn)
+            if c is not None:
+                rtn.append(c)
+            else:
+                raise ValueError(f"Coil {cn} is not a current optimisable coil")
+        return rtn
+
+    @property
+    def _opt_currents_inds(self) -> list[int]:
+        """
+        Get the indices of the coils that can be optimised.
+
+        These indices are used to extract the optimisable currents from the CoilSet
+        and are based on the index of the coils in the name array.
+        """
+        return [self.name.index(cn) for cn in self.current_optimisable_coil_names]
+
+    @property
+    def _opt_currents_symmetry_status(self) -> CoilSetSymmetryStatus:
+        """
+        Get the symmetry status of the CoilSet for current optimisations.
+
+        Notes
+        -----
+            For FULL and NONE symmetry status, analytic derivatives can be used.
+            When the status is FULL, the derivative values must be halved
+            after applying the repetition matrix as they will be added together.
+            For PARTIAL symmetry status, numerical derivatives must be used.
+        """
+        if all(isinstance(c, SymmetricCircuit) for c in self._coils):
+            return CoilSetSymmetryStatus.FULL
+        if any(isinstance(c, SymmetricCircuit) for c in self._coils):
+            return CoilSetSymmetryStatus.PARTIAL
+        return CoilSetSymmetryStatus.NONE
+
+    @property
+    def _opt_currents_expand_mat(self) -> np.ndarray:
+        """
+        Get the optimisation currents expansion matrix.
+
+        This matrix is used to convert the optimisable currents to the full set of
+        currents in the CoilSet.
+        """
+        cc = self.get_control_coils()
+
+        n_all_coils = cc.n_coils()
+        n_opt_coils = cc.n_current_optimisable_coils
+        n_distinct_coils_and_groupings = len(cc._coils)
+
+        if cc._opt_currents_symmetry_status == CoilSetSymmetryStatus.NONE:
+            return np.eye(n_all_coils)
+
+        # this should be true as, at the top level, the number
+        # of coil or group objects should be the same as the no
+        # of optimisable coils
+        assert n_opt_coils == n_distinct_coils_and_groupings  # noqa: S101
```

Comment:
Hmm, last time I checked (and it was a while ago) there should be no asserts in production code. Is this necessary now that development on this branch is closing up?

Reply:
That is true.

There some aasserts in BM. They are usually optimsed away and only enforced during development but I'm not entirly sure how that works in python (as there is no build step).

I'll probably just convert it into a raise but not really an error, it's more like logically these two values must be equal.